### DISCUSSION TITLE
Negative keyword operator at search

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#181] [Feedback & Feature Proposal]
The feature proposal is to introduce an operator `-` to exclude terms from the documents to be matched.

We could use it in this way to exclude the term `cars` from the documents to be returned at search time.

```json
{
    "q": "-cars"
}
```

**Comment by berti92** [comment]
As discussed with Olu on discord I should write my use case for this feature:

I implemented Meilisearch as a replacement for Elasticsearch for my document management system. I do not have a detailed filter function in my software, because 99% of the searches were possible by searching with the right syntax.

For example: "application -2019". Search every document with the string "application" but not documents which includes the string "2019".

I know that Meilisearch is not a drop in replacement for elasticsearch and that your software has not the same featureset, but I think this would be a good addition.

**Reply by gmourier** [reply]
Hello @berti92

Thanks for the effort in writing it here.

Unfortunately, I don't see a workaround in the meantime.

I understand the point of having this kind of expression and it could be a complement to the [phrase search](https://docs.meilisearch.com/learn/what_is_meilisearch/features.html#synonyms) expression.

I have a couple of questions that come to mind: when you type `-term`, do you expect the term to be exactly written the same way in the document?

By this, I mean having no matching via synonyms or typo tolerance corrections to the query term being negated.

In addition, the search would be limited to a single term in the form presented here. Should it be compatible with multiple terms? For example, `fire service -"multiple terms"`.

If yes, do you expect to exclude documents containing the terms in that order or could it also negate documents containing `terms multiple` ?

As you can see from the discussion, this is a need that seems to have little enthusiasm, but having discussed it in the past with the team, it has been identified as a low-hanging fruit in its instinctive form (`-term`), so perhaps we could make a good first question and facilitate external input on this topic if we don't prioritize it over other more important topics.

**Comment by macraig** [comment]
Hi everyone, we're happy to share that we've started working on adding negative keyword operators. It's planned as part of our v1.8 release, coming out on May 6th.

**Comment by macraig** [comment]
Hello everyone 👋

We have just released the first RC (release candidate) of Meilisearch containing this new feature!
You can test it by using

- the [release assets](https://github.com/meilisearch/meilisearch/releases)
- the [Meilisearch Docker image](https://hub.docker.com/repository/docker/getmeili/meilisearch/general)

```
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:v1.8.0-rc.0

```

You are welcome to leave your feedback in this discussion.

If you encounter any bugs, please report them [here](https://github.com/meilisearch/meilisearch/issues/new?assignees=&labels=&template=bug_report.md&title=).
Thanks in advance for your help and your involvement in Meilisearch ❤️

🎉 Official and stable release containing this change will be available on May 6th, 2024

⚠️ RC (release candidates) are not recommended for production
